### PR TITLE
Upgrade twilio libs

### PR DESCRIPTION
**Pull request by joliveros**
- Upgrade twilio libs
 - Replace Qunit w/ karma and mocha

### CODE REVIEW

File: public/js/tests/tests.js [hunk 1]
```diff
@@ -1,90 +1,79 @@
-test("should sort channels by name", function(assert) {
-  var channels = [
-    {friendlyName: "BBB"},
-    {friendlyName: "BBA"}
-  ];
+document.body.innerHTML = window.__html__['views/index.jade'];
 
-  var result = twiliochat.sortChannelsByName(channels)
+describe('TwilioChat', function() {
+  it('should sort channels by name', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: As a recomendation we'd like to write the description of each test like this:

```javascript
it("sorts channels by name", function() {
  // ...
```

The _should_ word is implicit in the context of a test.

When you read the test's output you'll find it more understandable IMO, what do you think?

File: public/js/tests/tests.js [hunk 2]
```diff
@@ -1,90 +1,79 @@
...
+      {friendlyName: 'BBB'},
+      {friendlyName: 'BBA'}
+    ];
 
-  assert.deepEqual(result, [{friendlyName: "BBA"}, {friendlyName: "BBB"}]);
-});
-
-test("should be able to sort an empty list of channels", function(assert) {
-  var channels = [];
+    var result = twiliochat.sortChannelsByName(channels)
```
Comment (acamino) [inline public/js/tests/tests.js]: Are we missing a semicolon?

File: public/js/tests/tests.js (hunk 1 above)
Comment (joliveros) [inline public/js/tests/tests.js]: yes, absolutely agree

File: public/js/tests/tests.js [hunk 3]
```diff
@@ -1,90 +1,79 @@
...
 
-  var result = twiliochat.sortChannelsByName(channels)
+    assert.deepEqual(result, [{friendlyName: 'BBA'}, {friendlyName: 'BBB'}]);
+  });
 
-  assert.deepEqual(result, []);
-});
+  it('should be able to sort an empty list of channels', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: This example makes me think we're missing a context - looks like we're trying to describe the function `sortChannelsByName`.

```javascript
describe("sortChannelsbyName", function() {
  it("sorts channels by name", function() {
    // ...
  });

  it("sorts an empty list", function() {
    // This case is good, since an empty list
    // is the obvious edge case for a list.
  });
});

File: public/js/tests/tests.js [hunk 4]
```diff
@@ -1,90 +1,79 @@
...
-  ];
+    var result = twiliochat.sortChannelsByName(channels)
 
-  var result = twiliochat.sortChannelsByName(channels);
+    assert.deepEqual(result, []);
+  });
 
-  assert.deepEqual(result, [{friendlyName: "BBA"}, {friendlyName: "BBA"}, {friendlyName: "BBB"}]);
-});
+  it('should sort channels when they have same name', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: This case is already covered by _sorts channels by name_. What about replacing the content of the first case with this content?

File: public/js/tests/tests.js [hunk 5]
```diff
@@ -1,90 +1,79 @@
...
- twiliochat.messagingClient = messagingClientMock;
- mock.expects("createChannel").once().returns({ then: function(){} });
- twiliochat.generalChannel = undefined;
- twiliochat.joinGeneralChannel();
+  it('should be able to add messages to chat', function() {
+    var message = {
+      body: 'just a test message',
+      author: 'me',
+      timestamp: new Date()
+    }
```
Comment (acamino) [inline public/js/tests/tests.js]: Missing semicolon here.

File: public/js/tests/tests.js [hunk 6]
```diff
@@ -1,90 +1,79 @@
...
+    var messageList = twiliochat.$messageList;
+    twiliochat.addMessageToList(message);
 
- mock.verify();
- ok(true);
-});
+    assert.ok(messageList.html().indexOf('just a test message') > -1, messageList.html());
```
Comment (acamino) [inline public/js/tests/tests.js]: Not sure why the original author introduced `.html()`. Does it make sense to you?

Comment (acamino) [inline public/js/tests/tests.js]: What do you think about using `isOk` instead of `ok`?

File: public/js/tests/tests.js [hunk 7]
```diff
@@ -1,90 +1,79 @@
...
+  });
 
-test("should not create a new general channel if it already has one", function(){
- var messagingClientMock = { createChannel: function () {} };
- var mock = sinon.mock(messagingClientMock);
- twiliochat.messagingClient = messagingClientMock;
- mock.expects("createChannel").never().returns({then: function(){} });
- twiliochat.generalChannel = {join: function(){ return {then: function() {}}}};
- twiliochat.joinGeneralChannel();
+  it('should create a general channel when there is not one', function(){
```
Comment (acamino) [inline public/js/tests/tests.js]: This description is difficult to understand. What about "creates a general channel if not present".

File: public/js/tests/tests.js [hunk 8]
```diff
@@ -1,90 +1,79 @@
...
-});
+   mock.verify();
+  });
 
-test("should create a new channel when requested by the user", function(){
-  var messagingClientMock = { createChannel: function () {} };
-  var mock = sinon.mock(messagingClientMock);
-  twiliochat.messagingClient = messagingClientMock;
-  mock.expects("createChannel").once().returns({ then: function(){} });
+  it('should create a new channel when requested by the user', function(){
```
Comment (acamino) [inline public/js/tests/tests.js]: This should be "creates a new channel". The _when_ clause is redundant, right?

File: public/js/tests/tests.js [hunk 9]
```diff
@@ -1,90 +1,79 @@
...
-});
+    mock.verify();
+  });
 
-test("should retrieve list of channels", function() {
-  var messagingClientMock = {getChannels: function(){} };
-  var mock = sinon.mock(messagingClientMock);
-  twiliochat.messagingClient = messagingClientMock;
-  mock.expects("getChannels").once().returns({then: function(){} });
+  it('should retrieve list of channels', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: What about "gets a list of channels"?

File: public/js/twiliochat.js [hunk 10]
```diff
@@ -90,12 +90,15 @@ var twiliochat = (function() {
   function connectMessagingClient(tokenResponse) {
```
Comment (acamino) [inline public/js/twiliochat.js]: What about just passing the actual token instead of the object which contains it?

File: public/js/twiliochat.js [hunk 11]
```diff
@@ -90,12 +90,15 @@ var twiliochat = (function() {
...
     // Initialize the IP messaging client
     tc.accessManager = new Twilio.AccessManager(tokenResponse.token);
```
Comment (acamino) [inline public/js/twiliochat.js]: This one could will be:

```javascript
 tc.accessManager = new Twilio.AccessManager(token); // or even better - "capabilityToken"
```

File: views/index.jade [hunk 12]
```diff
@@ -71,6 +71,7 @@ html(lang='en')
     script(src="/js/vendor/jquery.loadTemplate-1.4.4.min.js")
     script(src='//maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js')
     script(src='//cdnjs.cloudflare.com/ajax/libs/moment.js/2.11.1/moment.min.js')
-    script(src='//media.twiliocdn.com/sdk/rtc/js/ip-messaging/v0.9/twilio-ip-messaging.min.js')
+    script(src="//media.twiliocdn.com/sdk/js/common/releases/0.1.5/twilio-common.js")
+    script(src="//media.twiliocdn.com/sdk/js/chat/releases/0.11.1/twilio-chat.js")
```
Comment (acamino) [inline views/index.jade]: Cool!

File: public/js/tests/tests.js [hunk 13]
```diff
@@ -1,90 +1,85 @@
...
+      var channels = [];
+
+      var result = twiliochat.sortChannelsByName(channels);
+
+      assert.deepEqual(result, []);
+    });
+  });
+
+  describe('messaging', function() {
+    it('be able to add messages to chat', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: _It be able ..._ doesn't make sense, right?
Please update the description to "adds a message to chat" - the descriptions should be readable as plain english.

File: public/js/tests/tests.js [hunk 14]
```diff
@@ -1,90 +1,85 @@
...
+
+      var result = twiliochat.sortChannelsByName(channels);
+
+      assert.deepEqual(result,
+                       [{friendlyName: 'BBA'}, {friendlyName: 'BBA'},
+                       {friendlyName: 'BBB'}]
+      );
+    });
+
+    it('be able to sort an empty list of channels', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: This one should be "sorts an empty list" - of channels is inferred from its execution context. Would you mind changing it please?

File: public/js/twiliochat.js [hunk 15]
```diff
@@ -150,40 +154,60 @@ var twiliochat = (function() {
...
-      channel.on('messageAdded', tc.addMessageToList);
-      channel.on('typingStarted', showTypingStarted);
-      channel.on('typingEnded', hideTypingStarted);
-      channel.on('memberJoined', notifyMemberJoined);
-      channel.on('memberLeft', notifyMemberLeft);
-      $inputText.prop('disabled', false).focus();
-      tc.$messageList.text('');
-    });
+    return leaveCurrentChannel()
+      .then(initChannel(channel)
```
Comment (joliveros) [inline public/js/twiliochat.js]: I would avoid nesting chains of promises when possible:
```
function setupChannel(channel) {
    return leaveCurrentChannel()
      .then(initChannel(channel))
      .then(function(_channel) {
          return joinChannel(_channel); // you need the return here, otherwise the next promise will run too soon
      })
      .then(initChannelEvents);
  }
```

File: public/js/twiliochat.js [hunk 16]
```diff
@@ -150,40 +154,60 @@ var twiliochat = (function() {
...
+      .then(function(joinedChannel) {
+        console.log('Joined channel ' + joinedChannel.friendlyName);
+        updateChannelUI(_channel);
+        tc.currentChannel = _channel;
+        tc.loadMessages();
+        return joinedChannel;
+      });
+  }
+
+  function initChannelEvents() {
```
Comment (joliveros) [inline public/js/twiliochat.js]: +1

File: public/js/tests/tests.js [hunk 17]
```diff
@@ -1,90 +1,85 @@
...
+      var messageList = twiliochat.$messageList;
+      twiliochat.addMessageToList(message);
+
+      assert.isOk(messageList.html().indexOf('just a test message') > -1,
+                  messageList.html()
+      );
+    });
+  });
+
+  describe('channels creation', function() {
```
Comment (joliveros) [inline public/js/tests/tests.js]: I believe this suite is focused on `joinGeneralChannel()`

File: public/js/tests/tests.js [hunk 18]
```diff
@@ -1,90 +1,85 @@
...
+    it('be able to sort an empty list of channels', function() {
+      var channels = [];
+
+      var result = twiliochat.sortChannelsByName(channels);
+
+      assert.deepEqual(result, []);
+    });
+  });
+
+  describe('messaging', function() {
```
Comment (joliveros) [inline public/js/tests/tests.js]: I believe this suite is focused on `addMessageToList(message)`

File: public/js/tests/tests.js [hunk 19]
```diff
@@ -1,90 +1,85 @@
...
-  var messagingClientMock = {getChannels: function(){} };
-  var mock = sinon.mock(messagingClientMock);
-  twiliochat.messagingClient = messagingClientMock;
-  mock.expects("getChannels").once().returns({then: function(){} });
-
-  twiliochat.loadChannelList();
-
-  mock.verify();
-  ok(true);
+// This line transpiles the *.jade to HTML,via Pug,sets it as the document.body.
```
Comment (acamino) [inline public/js/tests/tests.js]: Whenever you have to add a comment like this you might want to extract a variable which explain what you put in the comment, e.g.

```javascript
var indexPageAsHtml = window.__html__["views/index.jade"];
document.body.innerHTML = indexPageAsHtml;
```

Then, you can get rid of that comment. The code speaks by itself. What do you think?

File: public/js/twiliochat.js (hunk 10 above)
Comment (acamino) [inline public/js/twiliochat.js]: This method is called from `fetchAccessToken`, right?

What about refactoring `fetchAccessToken` into something like this:

```javascript
function fetchAccessToken(username, handler) {
  $.post("/token", { identity: username, device: "browser" }, null, "json")
    .done(function(resp) {
      handler(resp.token); // We pass only the "token" to "connectMessagingClient".
    });
}
```

`connectMessagingClient` doesn't need to know about the inner structure of the object that contains the token which is the benefit of doing this refactor. I'd like to hear your thoughts on this.

File: public/js/tests/tests.js [hunk 20]
```diff
@@ -1,90 +1,85 @@
...
-
-  twiliochat.loadChannelList();
-
-  mock.verify();
-  ok(true);
+var indexPageAsHtml = window.__html__['views/index.jade'];
+document.body.innerHTML = indexPageAsHtml;
+
+describe('TwilioChat', function() {
+  context('channels sorting', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: This context should be `#sortChannelsByName()`, right?

File: public/js/tests/tests.js [hunk 21]
```diff
@@ -1,90 +1,85 @@
...
+    it('sorts an empty list', function() {
+      var channels = [];
+
+      var result = twiliochat.sortChannelsByName(channels);
+
+      assert.deepEqual(result, []);
+    });
+  });
+
+  context('#addMessageToList(message)', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: Usually we don't specify the argument on a _context_, e.g.:

```javascript
context('#addMessageToList()', function() {
  // ...
```

File: public/js/tests/tests.js [hunk 22]
```diff
@@ -1,90 +1,85 @@
...
+    it('adds a message to chat', function() {
+      var message = {
+        body: 'just a test message',
+        author: 'me',
+        timestamp: new Date(),
+      };
+      var messageList = twiliochat.$messageList;
+      twiliochat.addMessageToList(message);
+
+      assert.isOk(messageList.html().indexOf('just a test message') > -1,
```
Comment (acamino) [inline public/js/tests/tests.js]: Why to we call the method `.html()`. Is there any benefit on doing it?

File: public/js/tests/tests.js [hunk 23]
```diff
@@ -1,90 +1,85 @@
...
-  twiliochat.loadChannelList();
-
-  mock.verify();
-  ok(true);
+var indexPageAsHtml = window.__html__['views/index.jade'];
+document.body.innerHTML = indexPageAsHtml;
+
+describe('TwilioChat', function() {
+  context('channels sorting', function() {
+    it('sort channels by name', function() {
```
Comment (acamino) [inline public/js/tests/tests.js]: This one should be `sorts`. It reads as "it sorts channels by name".

**Comment by acamino** [comment]
@juancarlospaco you're almost there. Thanks.

**Comment by acamino** [comment]
Thanks! LGTM.
Please squash the commits, wait until the new (squashed) commit is green, then merge it.
